### PR TITLE
Fix edge cases around the websocket lifecycle.

### PR DESCRIPTION
**Pull request by charlesmchen**
Found even more issues around the websocket lifecycle.

PTAL @michaelkirk

### CODE REVIEW

File: src/Network/WebSockets/TSSocketManager.m [hunk 1]
```diff
@@ -57,7 +57,8 @@ - (instancetype)init
     OWSAssert([NSThread isMainThread]);
 
     _signalService = [OWSSignalService new];
-    
+    _status = kSocketStatusClosed;
```
Comment (charlesmchen) [inline src/Network/WebSockets/TSSocketManager.m:60]: We were neglecting to set this to a valid initial value.

File: src/Network/WebSockets/TSSocketManager.m [hunk 2]
```diff
@@ -117,6 +118,13 @@ - (void)ensureWebsocket
     DDLogWarn(@"Creating new websocket");
 
     // If socket is not already open or connecting, connect now.
+    //
+    // First we need to close the existing websocket, if any.
+    // The websocket methods are invoked _after_ the websocket state
+    // changes, so we may be just learning about a socket failure
+    // or close event now.
+    self.status = kSocketStatusClosed;
```
Comment (charlesmchen) [inline src/Network/WebSockets/TSSocketManager.m]: This is a key fix, as it ensures that we clear out any old failed/closed web sockets.

File: src/Network/WebSockets/TSSocketManager.m [hunk 3]
```diff
@@ -152,18 +160,20 @@ - (void)setStatus:(SocketStatus)status {
...
             case kSocketStatusClosed:
                 OWSAssert(!self.websocket);
                 break;
             case kSocketStatusOpen:
                 OWSAssert(self.websocket);
-                OWSAssert([self.websocket readyState] == SR_OPEN);
                 break;
             case kSocketStatusConnecting:
                 OWSAssert(self.websocket);
-                OWSAssert([self.websocket readyState] == SR_CONNECTING);
```
Comment (charlesmchen) [inline src/Network/WebSockets/TSSocketManager.m:166]: This wasn't safe - see comment above.

File: src/Network/WebSockets/TSSocketManager.m [hunk 4]
```diff
@@ -211,7 +221,13 @@ - (void)setStatus:(SocketStatus)status {
             socket.delegate = self;
             
             [self setWebsocket:socket];
+
+            // [SRWebSocket open] could hypothetically call a delegate method (e.g. if
+            // the socket failed immediately for some reason), so we update the status
+            // _before_ calling it, not after.
+            _status = status;
             [socket open];
+            return;
```
Comment (charlesmchen) [inline src/Network/WebSockets/TSSocketManager.m:230]: I don't think this is necessary in practice, but this isn't the first time we've hit edge cases with our web socket, so I'm trying to be super defensive.

File: src/Network/WebSockets/TSSocketManager.m [hunk 5]
```diff
@@ -250,12 +266,22 @@ - (void)closeWebSocket
 
 - (void)webSocketDidOpen:(SRWebSocket *)webSocket {
     OWSAssert([NSThread isMainThread]);
-    
+    OWSAssert(webSocket);
+    if (webSocket != self.websocket) {
+        // Ignore events from obsolete web sockets.
+        return;
+    }
```
Comment (charlesmchen) [inline src/Network/WebSockets/TSSocketManager.m:273]: This (and the corresponding changes below) does turn out to be necessary.  Apparently the websocket captures it's delegate _before_ dispatching the delegate method invocation.

**Comment by charlesmchen** [comment]
The repro around these issues is just to launch while the phone has no valid internet connection.

File: src/Network/WebSockets/TSSocketManager.m [hunk 6]
```diff
@@ -117,6 +118,13 @@ - (void)ensureWebsocket
     DDLogWarn(@"Creating new websocket");
 
     // If socket is not already open or connecting, connect now.
+    //
+    // First we need to close the existing websocket, if any.
+    // The websocket methods are invoked _after_ the websocket state
```
Comment (michaelkirk) [inline src/Network/WebSockets/TSSocketManager.m]: By "the websocket methods" do you mean "the websocket delegate methods"?

**Review by michaelkirk — APPROVED** [review]
LGTM

File: src/Network/WebSockets/TSSocketManager.m (hunk 6 above)
Comment (charlesmchen) [reply src/Network/WebSockets/TSSocketManager.m]: Done.

**Pull request merged**